client/storage.c

```diff
@@ -193,6 +193,304 @@ static int sort_callback(const void *l, const void *r)
+	item->size = size;
+	
+	/* strcpy(item->filename, filename); */
+	
+	item->next = NULL;
+	
+	pthread_mutex_lock(&simple_queue_mutex);
+	
+	uint8_t res = 0;
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		/* check if queue is ready for write. */
+		if(g_msg_queue[i].totalsize + size >= SIMPLE_QUEUE_MAX_MEM) {
+			g_msg_queue[i].status = !QUEUE_WRITE;
+		}
+		
+		if(g_msg_queue[i].status & QUEUE_WRITE) {
+			if(!g_msg_queue[i].head) {
+				g_msg_queue[i].head = item;
+				g_msg_queue[i].tail = item;
+			} else {
+				g_msg_queue[i].tail->next = item;
+				g_msg_queue[i].tail = item;
+			}
+			g_msg_queue[i].totalsize += item->size;
+			res = 1;
+			break;
+		}
+	}
+	
+	pthread_mutex_unlock(&simple_queue_mutex);
+	
+	if(res == 0) {
+		free(item);
+	}
+	
+	return res;
+}
+
+static struct queue_item* simple_queue_splice()
+{
+	pthread_mutex_lock(&simple_queue_mutex);
+	
+	struct queue_item * res = NULL;
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		/* check if queue is ready for write. */
+		if(g_msg_queue[i].head == NULL) {
+			g_msg_queue[i].status |= QUEUE_WRITE;
+			g_msg_queue[i].totalsize = 0;
+		} else if(!(g_msg_queue[i].status & QUEUE_WRITE)) {
+			res = g_msg_queue[i].head;
+			g_msg_queue[i].head = res->next;
+			g_msg_queue[i].totalsize -= res->size;
+			
+			if(g_msg_queue[i].tail == res) {
+				g_msg_queue[i].tail = NULL;
+				g_msg_queue[i].status |= QUEUE_WRITE;
+				g_msg_queue[i].totalsize = 0;
+			}
+			break;
+		}
+	}
+
+	pthread_mutex_unlock(&simple_queue_mutex);
+	return res;
+}
+
+struct init_storage_param {
+	xdag_time_t start_time; 
+	xdag_time_t end_time;
+	xdag_time_t *data;
+	void *(*callback)(void *, void *);
+};
+
+static void *init_storage_load_thread(void *data)
+{   
+	struct init_storage_param *param = (struct init_storage_param *)data;
+	xdag_time_t start_time = param->start_time;
+	xdag_time_t end_time = param->end_time;
+	
+	struct xdag_block readbuf[bufsize];
+	struct xdag_storage_sum s;
+	s.size = s.sum = 0;
+	
+	char path[256];
+	uint64_t pos = 0, mask;
+	int64_t todo = 0;
+	
+	while (start_time < end_time) {
+		sprintf(path, STORAGE_FILE, STORAGE_FILE_ARGS(start_time));
+				
+		pthread_mutex_lock(&storage_mutex);
+		FILE *f = xdag_open_file(path, "rb");
+		if (f) {
+			if (fseek(f, pos, SEEK_SET) < 0) todo = 0;
+			else todo = fread(readbuf, sizeof(struct xdag_block), bufsize, f);
+			xdag_close_file(f);
+		} else {
+			todo = 0;
+		}
+		pthread_mutex_unlock(&storage_mutex);
+		
+		if(todo > 0) {
+			for (int i = 0; i < todo; ++i) {
+				s.size += sizeof(struct xdag_block);
+				for (int j = 0; j < sizeof(struct xdag_block) / sizeof(uint64_t); ++j) {
+					s.sum += ((uint64_t*)(readbuf + i))[j];
+				}
+			}
+			
+			struct xdag_block *bs = (struct xdag_block*)malloc(sizeof(struct xdag_block)*todo);
+			memcpy(bs, readbuf, sizeof(struct xdag_block)*todo);
+			
+			while (simple_queue_push(bs, sizeof(struct xdag_block)*todo/*, path*/) == 0) {
+#if defined (_WIN32) || defined (_WIN64)
+				sleep(1);
```

Comment:
Why sleep for one second?

Reply:
It should be 1ms. My mistake for windows.